<?xml version="1.0"?>

<launch>

  <arg name="node_start_delay" default="5.0"/>

  <param name="use_sim_time" value="true"/>

  <remap from="scan" to="base_scan"/>

  <node pkg="planners" type="people_remap" name="people_remap">
    <remap from="people" to="/robot_0/move_base/RRTPlanner/people"/>
  </node>

  <group ns="robot_0">
    <!--param name="tf_prefix" value="robot_0"/-->
    <!--node pkg="planners" type="waypoint_markers" name="waypoint_markers">
      <param name="num_markers" value="7"/>
      <param name="base_frame" value="/robot_0/base_footprint"/>
      <param name="goal_tolerance" value="5.0"/>
    </node-->

    <!--node pkg="planners" type="obstacle_distance" name="obstacle_distance" output="screen">
      <param name="costmap_topic" value="move_base/local_costmap/costmap"/>
      <param name="base_frame" value="/robot_0/base_footprint"/>
    </node-->

    <node pkg="amcl" type="amcl" name="amcl">
      <param name="use_map_topic" value="true"/>
      <param name="base_frame_id" value="robot_0/base_footprint"/>
      <param name="odom_frame_id" value="robot_0/odom"/>
      <param name="tf_broadcast" value="true"/>
      <param name="odom_model_type" value="diff-corrected"/>
      <param name="odom_alpha1" value="0.0075"/>
      <param name="odom_alpha2" value="0.0075"/>
      <param name="odom_alpha3" value="0.0075"/>
      <param name="odom_alpha4" value="0.0075"/>
      <remap from="scan" to="base_scan"/>
      <remap from="static_map" to="/static_map"/>
      <remap from="map" to="/map"/>
    </node>
    <node pkg="move_base" type="move_base" name="move_base" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@'">
      <rosparam file="$(find planners)/launch/config/rrtmp_dynamic_costmap_common_params.yaml" command="load" ns="global_costmap"/>
      <rosparam file="$(find planners)/launch/config/rrtmp_dynamic_costmap_common_params.yaml" command="load" ns="local_costmap"/>
      <rosparam file="$(find planners)/launch/config/rrtmp_dynamic_local_costmap_params.yaml" command="load"/>
      <rosparam file="$(find planners)/launch/config/rrtmp_dynamic_global_costmap_params.yaml" command="load"/>
      <rosparam file="$(find planners)/launch/config/dynamic_local_planner_params.yaml" command="load"/>
      <param name="base_local_planner" value="rrt_planner/RRTPlanner"/>
      <param name="planner_frequency" value="1"/>
      <param name="recovery_behavior_enabled" value="false"/>
      <remap from="/robot_0/move_base/RRTPlanner/robot_0/cmd_vel" to="cmd_vel_0"/>
      <remap from="map" to="/map"/>
    </node>
  </group>

  <group ns="robot_1">
    <node pkg="planners" type="dynamic_obstacles" name="dynamic_obstacles">
      <remap from="/robot_1/dynamic_obstacles/cmd_vel" to="cmd_vel"/>
      <param name="obstacle_velocity" value="0.25"/>
      <param name="obstacle_steps" value="180"/>
      <param name="obstacle_delay_1" value="0"/>
      <param name="obstacle_delay_2" value="0"/>
    </node>
  </group>

  <group ns="robot_2">
    <node pkg="planners" type="dynamic_obstacles" name="dynamic_obstacles">
      <remap from="/robot_1/dynamic_obstacles/cmd_vel" to="cmd_vel"/>
      <param name="obstacle_velocity" value="0.25"/>
      <param name="obstacle_steps" value="180"/>
      <param name="obstacle_delay_1" value="1000000"/>
      <param name="obstacle_delay_2" value="7500"/>
    </node>
  </group>

  <group ns="robot_3">
    <node pkg="planners" type="dynamic_obstacles" name="dynamic_obstacles">
      <remap from="/robot_1/dynamic_obstacles/cmd_vel" to="cmd_vel"/>
      <param name="obstacle_velocity" value="0.25"/>
      <param name="obstacle_steps" value="180"/>
      <param name="obstacle_delay_1" value="1000000"/>
      <param name="obstacle_delay_2" value="7500"/>
    </node>
  </group>

  <group ns="robot_4">
    <node pkg="planners" type="dynamic_obstacles" name="dynamic_obstacles">
      <remap from="/robot_1/dynamic_obstacles/cmd_vel" to="cmd_vel"/>
      <param name="obstacle_velocity" value="0.25"/>
      <param name="obstacle_steps" value="200"/>
      <param name="obstacle_delay_1" value="1000000"/>
      <param name="obstacle_delay_2" value="17500"/>
    </node>
  </group>

  <group ns="robot_5">
    <node pkg="planners" type="dynamic_obstacles" name="dynamic_obstacles">
      <remap from="/robot_1/dynamic_obstacles/cmd_vel" to="cmd_vel"/>
      <param name="obstacle_velocity" value="0.25"/>
      <param name="obstacle_steps" value="200"/>
      <param name="obstacle_delay_1" value="1000000"/>
      <param name="obstacle_delay_2" value="17500"/>
    </node>
  </group>

  <group ns="robot_6">
    <node pkg="planners" type="dynamic_obstacles" name="dynamic_obstacles">
      <remap from="/robot_1/dynamic_obstacles/cmd_vel" to="cmd_vel"/>
      <param name="obstacle_velocity" value="0.25"/>
      <param name="obstacle_steps" value="180"/>
      <param name="obstacle_delay_1" value="1000000"/>
      <param name="obstacle_delay_2" value="17500"/>
    </node>
  </group>

  <group ns="robot_7">
    <node pkg="planners" type="dynamic_obstacles" name="dynamic_obstacles">
      <remap from="/robot_1/dynamic_obstacles/cmd_vel" to="cmd_vel"/>
      <param name="obstacle_velocity" value="0.25"/>
      <param name="obstacle_steps" value="200"/>
      <param name="obstacle_delay_1" value="1000000"/>
      <param name="obstacle_delay_2" value="22500"/>
    </node>
  </group>

  <group ns="robot_8">
    <node pkg="planners" type="dynamic_obstacles" name="dynamic_obstacles">
      <remap from="/robot_1/dynamic_obstacles/cmd_vel" to="cmd_vel"/>
      <param name="obstacle_velocity" value="0.25"/>
      <param name="obstacle_steps" value="180"/>
      <param name="obstacle_delay_1" value="1000000"/>
      <param name="obstacle_delay_2" value="30000"/>
    </node>
  </group>

  <node pkg="stage_ros" type="stageros" name="stageros" args="-d $(find stage_simulation)/worlds/rrtdmp_5.world">
    <!--remap from="tf" to="tf_old"/-->
    <remap from="/robot_0/cmd_vel" to="/robot_0/cmd_vel_0"/>
  </node>
  <node pkg="map_server" name="map_server" type="map_server" args="$(find stage_simulation)/worlds/maps/rrtdmp_5.yaml"/>

  <node pkg="planners" type="remap_scan" name="remap_scan"/>
  <node pkg="planners" type="clear_scan" name="clear_scan">
    <param name="laser_frame_id" value="robot_0/base_laser_link"/>
    <param name="reduction_factor" value="0.1"/>
  </node>

  <node pkg="planners" type="initialize" name="initialize" output="screen"/>

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find planners)/launch/config/rviz/rrtmp.rviz">
    <remap from="initialpose" to="robot_0/initialpose"/><!--remap rviz initial position button-->
    <remap from="move_base_simple/goal" to="robot_0/move_base_simple/goal"/><!--remap rviz goal position button-->
  </node>

  <node pkg="rosbag" type="record" name="rrtmp_record_trajectory"
    args="robot_0/move_base/RRTPlanner/trajectory
          robot_0/move_base/RRTPlanner/time_cost"/>

</launch>
